src/p2p/ServiceQueue.ts

```diff
@@ -0,0 +1,768 @@
+  record.txremove = txRemove.sort((a, b) => a.txHash.localeCompare(b.txHash))
+
+  // we need to get the hash of the txlist after the txadd and txremove
+  // but we dont want to alter the txList, so we make a copy
+  const txListCopy = clone(txList)
+
+  for (const txadd of record.txadd) {
+    const { sign, ...txDataWithoutSign } = txadd.txData
+    sortedInsert(txListCopy, {
+      hash: txadd.hash,
+      tx: {
+        hash: txadd.hash,
+        txData: txDataWithoutSign,
+        type: txadd.type,
+        cycle: txadd.cycle,
+        ...(txadd.subQueueKey && { subQueueKey: txadd.subQueueKey }),
+      },
+    })
+  }
+
+  for (const txremove of record.txremove) {
+    const index = txListCopy.findIndex((entry) => entry.hash === txremove.txHash)
+    if (index === -1) {
+      error(`updateRecord: TxHash ${txremove.txHash} does not exist in txListCopy`)
+    } else {
+      txListCopy.splice(index, 1)
+    }
+  }
+
+  const reversedTxList = txListCopy.slice().reverse()
+  // add all active nodes to the cycle record in the event of a shutdown
+  if (record.mode === 'shutdown') {
+    for (const node of Nodelist.activeByIdOrder) {
+      if (record.activated.includes(node.id)) {
+        if (
+          record.txadd.some((entry) => entry.txData.nodeId === node.id && entry.type === 'nodeInitReward')
+        ) {
+          warn(
+            `shutdown condition: active node with id ${node.id} is already in txadd (nodeInitReward); this should not happen`
+          )
+        } else {
+          const txData = {
+            startTime: record.start,
+            publicKey: node.publicKey,
+            nodeId: node.id,
+          }
+
+          const hash = crypto.hash(txData)
+          const addTx: P2P.ServiceQueueTypes.AddNetworkTx = {
+            hash,
+            type: 'nodeInitReward',
+            txData: txData,
+            cycle: currentCycle,
+            subQueueKey: node.publicKey,
+          }
+
+          sortedInsert(txListCopy, {
+            hash: addTx.hash,
+            tx: {
+              hash: addTx.hash,
+              txData: txData,
+              type: addTx.type,
+              cycle: addTx.cycle,
+              ...(addTx.subQueueKey && { subQueueKey: addTx.subQueueKey }),
+            },
+          })
+
+          record.txadd.push(addTx)
+        }
+      }
+      if (record.txadd.some((entry) => entry.txData.nodeId === node.id && entry.type === 'nodeReward')) {
+        warn(`shutdown condition: active node with id ${node.id} is already in txadd; this should not happen`)
+        continue
+      }
+
+      // get latest entry for node in txList. and if it is init then we inject otherwise continue
+      // first iterate over txlist backwards and get first entry that has public key of node
+      const txListEntry = reversedTxList.find((entry) => entry.tx.txData.publicKey === node.publicKey)
+      if (txListEntry && txListEntry.tx.type !== 'nodeInitReward') {
```

Comment:
Taking this a techdebt.   We should replace with  app.getFinalNodeTXName()  an similar functions so that the final TX  names, shape and creation come from the app

Reply:
This is being refactored by Phillipp as this shouldn’t be in the shardus layer but rather in the app layer

Reply:
All the shutdown impl in update record is going to be changed when he pushes that tomorrow